ga4gh/datamodel/genotype_phenotype.py

```diff
@@ -0,0 +1,366 @@
+objects.
+"""
+
+import rdflib
+import urlparse
+import sets
+import os
+
+import ga4gh.protocol as protocol
+import ga4gh.exceptions as exceptions
+
+
+class G2PDataset:
+    """
+    An rdf object store.  The cancer genome database
+    [Clinical Genomics Knowledge Base]
+    (http://nif-crawler.neuinfo.org/monarch/ttl/cgd.ttl)
+    published by the Monarch project was the source of Evidence.
+    """
+
+    def __init__(self, setName, relativePath, dataDir):
+        """
+        Initialize dataset, using the passed dict of sources
+        [{source,format}] see rdflib.parse() for more
+        """
+        self._sources = [os.path.join(relativePath, f)
+                         for f in os.listdir(relativePath)]
+        self._searchQuery = """
+            PREFIX OBAN: <http://purl.org/oban/>
+            PREFIX OBO: <http://purl.obolibrary.org/obo/>
+            PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+            PREFIX faldo: <http://biohackathon.org/resource/faldo#>
+            PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+            SELECT %PROPERTIES%
+                WHERE {
+                    ?s    a OBAN:association .
+                    ?s  OBAN:association_has_subject ?l .
+                    ?l  rdfs:label ?location_label  .
+                    %LOCATION%
+                    ?s  OBO:RO_has_environment  ?drug .
+                    ?drug  rdfs:label ?drug_label  .
+                    ?s  OBAN:association_has_object  ?d .
+                    ?d  rdfs:label ?disease_label  .
+                    ?d rdf:type ?disease .
+                    ?s  OBAN:association_has_object_property  ?evidence .
+                    OPTIONAL {  ?evidence  rdfs:label ?evidence_label } .
+              %FILTER%
+            }
+            """
+
+        # initialize graph
+        self._rdfGraph = rdflib.ConjunctiveGraph()
+
+        # load with data
+        for source in self._sources:
+            if source.endswith('.ttl'):
+                self._rdfGraph.parse(source,
+                                     format='n3')
+            else:
+                self._rdfGraph.parse(source,
+                                     format='xml')
+
+        # log queries that take more than N seconds
+        # import time
+        # self.RESPONSETIME_LOGGING_THRESHOLD = 2
+
+    def _search(self, request):
+        offset = request.offset
+
+        # now = time.time()
+        associations = self.queryLabels(
+            request.feature, request.evidence, request.phenotype,
+            request.pageSize, offset)
+        # responseTime = time.time()-now
+        # if responseTime > self.RESPONSETIME_LOGGING_THRESHOLD:
+        #     print('_search', responseTime)
+        #     print(request)
+
+        self.associationsLength = len(associations)
+        for association in associations:
+            yield association
+
+    def queryLabels(
+         self, location=None, drug=None, disease=None, pageSize=None,
```

Comment:
This pageSize and offset stuff is horribly tricky. We can't just pass pageSize directly in to the LIMIT statement here because we have potential DOS attacks. We need to be able to limit the size of the output buffer, and this has to be done at the top level. The only way this works efficiently and cleanly for large result sets is if we can get a stream of objects from the rdflib graph, and be able to restart that stream from a particular offset. If we don't expect to stream large sets of results back, and realistically should be able to fit all results onto a single page, then we should forget about paging entirely at this level and just return a list of all the results.